designcompose/src/main/java/com/android/designcompose/squoosh/DESIGN.md

```diff
@@ -0,0 +1,92 @@
+- Lists and Grids is entirely unimplemented.
+
+## Modules
+Squoosh has a few key modules and functions:
+
+ - `SquooshResolvedNode` is a design node in a tree that can be sent to layout. It's "resolved"
+   because variants and some customizations have been applied.
+ - `resolveVariantsRecursively` generates a `SquooshResolvedNode` tree from a DesignCompose `View`
+   and all of the context (customizations, interaction state, etc). This function is currently
+   slow and needs some optimization.
+ - `updateLayoutTree` generates a `LayoutNodeList` from a `SquooshResolvedNode` and any previous
+   set of layout values. This is typically pretty fast.
+ - `SquooshLayout.doLayout` calls the JNI layout implementation to perform a layout iteration.
+ - `populateComputedLayout` iterates a `SquooshResolvedNode` tree and copies values from the JNI
+   layout iteration into the relevant `SquooshResolvedNode`.
+ - `mergeTreesAndCreateSquooshAnimationControl` inspects two `SquooshResolvedNode` trees and updates
+   the destination tree to have all of the nodes needed to show a transition from the "from" tree
+   into the "to" tree.
+ - `Modifier.squooshRender` knows how to render an entire `SquooshResolvedNode`--typically
+   corresponding to the entire Composable.
+
+## Typical Execution
+Whenever squoosh needs to render, it will:
+
+1. Build a new `SquooshResolvedNode` tree with `resolveVariantsRecursively`
+2. Create a request to JNI layout using `updateLayoutTree`
+3. Submit the request to JNI layout using `SquooshLayout.doLayout`
+4. Put the results of the layout into the `SquooshResolvedNode` tree so that rendering is easier
+   using `populateComputedLayout`
+5. If there are transitions, then generate a new `InteractionState` (and soon `CustomizationContext`)
+   with the transition values applied (these state objects start out without the action of the
+   transition applied, then we ask them to make a copy with all active transitions applied -- so it
+   is the state of the world once all transitions are complete). We then run through steps 1-4 again
+   with the updated context.
+    * This gives us a `SquooshResolvedNode` tree of the world after all transitions have ended.
+    * We use `newSquooshAnimate` to create `SquooshAnimationControl` objects that control progress
+      for each transition. `newSquooshAnimate` identifies similar nodes and different nodes and
+      creates objects that will animate the appropriate property for each. Some nodes from the source
+      tree get moved to the destination tree so that they can be faded out.
+6. If there are transitions, we launch an effect to run an animation clock and update the state of
+   each transition's animation. This is how we can seemingly run many different animations at once
+   -- they all just show different positions between the two trees, and can't be nested.
+7. We emit a `Layout` Composable that occupies the appropriate space and has a `squooshRender`
```

Comment:
What do you mean by emitting a `Layout` composable? Is this related to the custom layout function?

Reply:
I meant the basic [Compose Layout composable function](https://developer.android.com/reference/kotlin/androidx/compose/ui/layout/package-summary#Layout(kotlin.collections.List,androidx.compose.ui.Modifier,androidx.compose.ui.layout.MultiContentMeasurePolicy)). Currently Squoosh makes one of those that does drawing (via the `squoohRender` Modifier), and then some other child ones that handle input events and wrap nested Composables (via replacement, etc).

I think that Squoosh should probably be changed so that instead of the root Layout doing all rendering itself, the children also do some rendering. The disadvantage of moving to this approach is that it's harder to support multiple masks and things like that over multiple children.

Reply:
I updated the doc, it's a bit clearer now, and we can do more iteration on it in Docs.